ProjFS.Mac/PrjFSKext/VnodeCache.cpp

```diff
@@ -0,0 +1,313 @@
+    PerfTracer* _Nonnull perfTracer,
+    PrjFSPerfCounter cacheMissFallbackFunctionCounter,
+    PrjFSPerfCounter cacheMissFallbackFunctionInnerLoopCounter,
+    vfs_context_t _Nonnull context,
+    vnode_t _Nonnull vnode,
+    uintptr_t vnodeHash,
+    uint32_t vnodeVid,
+    bool invalidateEntry,
+    /* out parameters */
+    VirtualizationRootHandle& rootHandle)
+{
+    RWLock_AcquireExclusive(s_entriesLock);
+    {
+        if (!TryUpdateEntryToLatest_ExclusiveLocked(
+                perfTracer,
+                cacheMissFallbackFunctionCounter,
+                cacheMissFallbackFunctionInnerLoopCounter,
+                vnode,
+                context,
+                vnodeHash,
+                vnodeVid,
+                invalidateEntry,
+                /* out */ rootHandle))
+        {
+            // TryUpdateEntryToLatest_ExclusiveLocked can only fail if the cache is full
+            
+            perfTracer->IncrementCount(PrjFSPerfCounter_CacheFullCount, true /*ignoreSampling*/);
+        
+            InvalidateCache_ExclusiveLocked();
+            if (!TryUpdateEntryToLatest_ExclusiveLocked(
+                        perfTracer,
+                        cacheMissFallbackFunctionCounter,
+                        cacheMissFallbackFunctionInnerLoopCounter,
+                        vnode,
+                        context,
+                        vnodeHash,
+                        vnodeVid,
+                        true, // invalidateEntry
+                        /* out */ rootHandle))
+            {
+                KextLog_FileError(
+                    vnode,
+                    "ClearCacheAndInsertEntry: failed to insert vnode (%p:%u) after emptying cache",
+                    KextLog_Unslide(vnode), vnodeVid);
+                
+                rootHandle = VirtualizationRoot_FindForVnode(
+                    perfTracer,
+                    cacheMissFallbackFunctionCounter,
+                    cacheMissFallbackFunctionInnerLoopCounter,
+                    vnode,
+                    context);
+            }
+        }
+    }
+    RWLock_ReleaseExclusive(s_entriesLock);
+}
+
+KEXT_STATIC bool TryFindVnodeIndex_Locked(
+    vnode_t _Nonnull vnode,
+    uintptr_t vnodeHash,
+    /* out parameters */
+    uintptr_t& vnodeIndex)
+{
+    // Walk from the starting index until we do one of the following:
+    //    -> Find the vnode
+    //    -> Find where the vnode should be inserted (i.e. NULLVP)
+    //    -> Have looped all the way back to where we started
+    vnodeIndex = vnodeHash;
+    while (vnode != s_entries[vnodeIndex].vnode && NULLVP != s_entries[vnodeIndex].vnode)
+    {
+        vnodeIndex = (vnodeIndex + 1) % s_entriesCapacity;
+        
+        if (vnodeIndex == vnodeHash)
+        {
+            // Looped through the entire cache and didn't find an empty slot or the vnode
+            return false;
+        }
+    }
+    
+    return true;
+}
+
+KEXT_STATIC bool TryUpdateEntryToLatest_ExclusiveLocked(
```

Comment:
This looks like it assumes the Vnode exists, and if it doesn't it does nothing and returns false.
It's somewhat clear for sure but a comment above would help.